From 1da209df5742c2d4d4fd8f98b1925f1b0d565dce Mon Sep 17 00:00:00 2001
From: rany <ranygh@riseup.net>
Date: Fri, 12 May 2023 20:03:41 +0300
Subject: [PATCH] wifi: mt76: mt7921: add 256-QAM support in 2.4Ghz

Signed-off-by: rany <ranygh@riseup.net>
---
 mt7921/init.c | 6 ++++++
 1 file changed, 6 insertions(+)

--- a/mt7921/init.c
+++ b/mt7921/init.c
@@ -317,6 +317,12 @@ int mt7921_register_device(struct mt792x
 	dev->mphy.sband_5g.sband.ht_cap.cap |=
 			IEEE80211_HT_CAP_LDPC_CODING |
 			IEEE80211_HT_CAP_MAX_AMSDU;
+	dev->mphy.sband_2g.sband.vht_cap.cap |=
+			IEEE80211_VHT_CAP_MAX_MPDU_LENGTH_11454 |
+			IEEE80211_VHT_CAP_MAX_A_MPDU_LENGTH_EXPONENT_MASK |
+			IEEE80211_VHT_CAP_SU_BEAMFORMEE_CAPABLE |
+			IEEE80211_VHT_CAP_MU_BEAMFORMEE_CAPABLE |
+			(3 << IEEE80211_VHT_CAP_BEAMFORMEE_STS_SHIFT);
 	dev->mphy.sband_5g.sband.vht_cap.cap |=
 			IEEE80211_VHT_CAP_MAX_MPDU_LENGTH_11454 |
 			IEEE80211_VHT_CAP_MAX_A_MPDU_LENGTH_EXPONENT_MASK |
